### PR TITLE
Pantelides algorithm

### PR DESCRIPTION
**Pull request by YingboMa**

### CODE REVIEW

File: test/index_reduction.jl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        0 ~ x^2 + y^2 - L^2]
+pendulum2 = ODESystem(eqs2, t, [x, y, T], [L, g], name=:pendulum)
+lowered_sys = ModelingToolkit.ode_order_lowering(pendulum2)
+
+lowered_eqs = [D(x_t) ~ T*x,
+               D(y_t) ~ T*y - g,
+               0 ~ x^2 + y^2 - L^2,
+               D(x) ~ x_t,
+               D(y) ~ x_t]
+@test_skip ODESystem(lowered_eqs) == lowered_sys # not gonna work
```
Comment (ChrisRackauckas) [inline test/index_reduction.jl:22]: why?

File: src/systems/diffeqs/index_reduction.jl [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            end
+        end
+    end
+
+    fullvars = [xvars; dxvars; algvars] # full list of variables
+    vars_asso = [(1:xvar_offset) .+ xvar_offset; zeros(Int, length(fullvars) - xvar_offset)] # variable association list
+    return edges, fullvars, vars_asso
+end
+
+print_bigraph(sys, vars, edges) = print_bigraph(stdout, sys, vars, edges)
```
Comment (ChrisRackauckas) [inline src/systems/diffeqs/index_reduction.jl:54]: would there be an advantage to using LightGraphs in here?

Comment (YingboMa) [reply src/systems/diffeqs/index_reduction.jl:54]: I don't think so. The graph algorithms here are pretty self-contained. Also, `print_bigraph` is only used for debugging.

File: test/index_reduction.jl (hunk 1 above)
Comment (YingboMa) [reply test/index_reduction.jl:22]: We cannot differentiate between time-dependent parameters and algebraic variables. So, `ODESystem(lowered_eqs)` construction is not going to work.

Comment (ChrisRackauckas) [reply test/index_reduction.jl:22]: You shouldn't be using the DSL here.

Comment (ChrisRackauckas) [inline test/index_reduction.jl:22]:
```suggestion
@test ODESystem(lowered_eqs,[x,y,x_t,y_t,T],[L, g]) == lowered_sys
```

File: test/index_reduction.jl [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+@variables x(t) y(t) w(t) z(t) T(t) x_t(t) y_t(t)
+@derivatives D'~t
+
+eqs2 = [D(D(x)) ~ T*x,
+        D(D(y)) ~ T*y - g,
+        0 ~ x^2 + y^2 - L^2]
+pendulum2 = ODESystem(eqs2, t, [x, y, T], [L, g], name=:pendulum)
+lowered_sys = ModelingToolkit.ode_order_lowering(pendulum2)
+
+lowered_eqs = [D(x_t) ~ T*x,
```
Comment (ChrisRackauckas) [inline test/index_reduction.jl:17]: that's the wrong ˍ

**Pull request merged**